src/creation/mod.rs

```diff
@@ -0,0 +1,144 @@
+    loop {
+        if can_create {
+            if let Some(result) = creator.create_unit(round) {
+                return Ok(result);
+            }
+        }
+        futures::select! {
+            unit = incoming_parents.next() => match unit {
+                Some(unit) => creator.add_unit(&unit),
+                None => {
+                    info!(target: "AlephBFT-creator", "Incoming parent channel closed, exiting.");
+                    return Err(());
+                }
+            },
+            _ = &mut delay => {
+                if can_create {
+                    warn!(target: "AlephBFT-creator", "Delay passed despite us not waiting for it.");
+                }
+                can_create = true;
+                delay = Delay::new(Duration::from_secs(30 * 60)).fuse();
+            },
+            _ = exit => {
+                info!(target: "AlephBFT-creator", "Received exit signal.");
+                return Err(());
+            },
+        }
+    }
+}
+
+/// A process responsible for creating new units. It receives all the units added locally to the Dag
+/// via the `incoming_parents` channel. It creates units according to an internal strategy respecting
+/// always the following constraints: if round is equal to 0, U has no parents, otherwise for a unit U of round r > 0
+/// - all U's parents are from round (r-1),
+/// - all U's parents are created by different nodes,
+/// - one of U's parents is the (r-1)-round unit by U's creator,
+/// - U has > floor(2*N/3) parents.
+/// - U will appear in the channel only if all U's parents appeared there before
+/// The currently implemented strategy creates the unit U according to a delay schedule and when enough
+/// candidates for parents are available for all the above constraints to be satisfied.
+///
+/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/internals.html
+/// Section 5.1 for a discussion of this component.
+pub async fn run<H: Hasher>(
+    conf: Config,
+    io: IO<H>,
+    starting_round: oneshot::Receiver<Round>,
+    mut exit: oneshot::Receiver<()>,
+) {
+    let Config {
+        node_id,
+        n_members,
+        create_lag,
+        max_round,
+    } = conf;
+    let mut creator = Creator::new(node_id, n_members);
+    let IO {
+        mut incoming_parents,
+        outgoing_units,
+    } = io;
+    let starting_round = match starting_round.await {
+        Ok(round) => round,
+        Err(e) => {
+            error!(target: "AlephBFT-creator", "Starting round not provided: {}", e);
+            return;
+        }
+    };
+    debug!(target: "AlephBFT-creator", "Creator starting from round {}", starting_round);
+    for round in starting_round..max_round {
+        // Skip waiting if we are way behind.
+        // We have to be two rounds behind, otherwise someone could force us to skip all delays.
+        let ignore_delay = creator.current_round() > round + 2;
```

Comment:
This is correct now, but it seems to me that if it was `> round ` instead of `> round + 2` then it would be still fine?
If there exists a unit at round `round+1`, then there must be several honest nodes who created a unit at round `round`, thus by creating a unit at `round` we will not be the "fastest" honest node. And by a separate reasoning we show that the fastest honest node always waits the full delay, so it should be fine.

Reply:
Hm, I think you are right, changing. <_<"